src/lib/krb5/ccache/ccselect.c

```diff
@@ -142,32 +144,55 @@ krb5_cc_select(krb5_context context, krb5_principal server,
             return ret;
     }
 
+    ret = krb5_copy_principal(context, server, &srvcp);
+    if (ret)
+        return ret;
+
+    /* Try to use the fallback host realm for the server if there's no
+     * authoritative realm. */
+    if (krb5_is_referral_realm(&srvcp->realm)) {
+        ret = krb5_get_fallback_host_realm(context, &srvcp->data[1], &fbrealms);
```

Comment:
This crashes if there aren't two components in the principal.  Check the number of components and the name type before assuming there's a hostname in the second component.